In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv(r"D:\Data Science 2024\Study\Advertising.csv")
df.head()

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [2]:
X = df.drop('Sales',axis=1)
y = df['Sales']

### Use PolynomialFeatures

In [4]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [5]:
polynomial_converter = PolynomialFeatures(degree=2,include_bias=False)

In [6]:
polynomial_converter.fit(X)

PolynomialFeatures(include_bias=False)

In [7]:
poly_features = polynomial_converter.transform(X)

X.shape

In [9]:
X.iloc[0]

TV           230.1
Radio         37.8
Newspaper     69.2
Name: 0, dtype: float64

In [10]:
poly_features[0]

array([2.301000e+02, 3.780000e+01, 6.920000e+01, 5.294601e+04,
       8.697780e+03, 1.592292e+04, 1.428840e+03, 2.615760e+03,
       4.788640e+03])

* first 3 terms are the origial values 230.1, 37.8, 69.2
* 4th term 5.2946 ~ (230.1 ** 2) Polynomial term 
* 5th term 8.6977 ~ 230.1 * 37.8 Interaction term
* 6th term 1.5922 ~ 230.1 * 69.2 Interaction term
* 7th term 1.4288 ~ (37.8 ** 2) Polynomial term
* 8th term 2.6157 ~ 37.8 * 69.2 Interaction term
* 9th term 4.7886 ~ (69.2 ** 2) Polynomial term

## As fit and Transform operations are mostly performed one after another and that's why one combined effect function is also available
## fit_transform(X) - Note : Nothing return until you use transform function

In [13]:
polynomial_converter.fit_transform(X)

array([[ 230.1 ,   37.8 ,   69.2 , ..., 1428.84, 2615.76, 4788.64],
       [  44.5 ,   39.3 ,   45.1 , ..., 1544.49, 1772.43, 2034.01],
       [  17.2 ,   45.9 ,   69.3 , ..., 2106.81, 3180.87, 4802.49],
       ...,
       [ 177.  ,    9.3 ,    6.4 , ...,   86.49,   59.52,   40.96],
       [ 283.6 ,   42.  ,   66.2 , ..., 1764.  , 2780.4 , 4382.44],
       [ 232.1 ,    8.6 ,    8.7 , ...,   73.96,   74.82,   75.69]])

### And the remaining steps are same as linear model, except we will be using Polynomial_features instead of X in train-test-split

In [15]:
X_train, X_test, y_train, y_test = train_test_split(poly_features,y,test_size=0.33, random_state=102)

In [16]:
model = LinearRegression()

In [17]:
model.fit(X_train,y_train)

LinearRegression()

In [18]:
test_predictions = model.predict(X_test)

In [19]:
test_predictions

array([19.978015  , 13.43579987, 15.31639164, 16.57814396,  9.98297337,
       20.06348368,  9.32150033,  6.85065356,  5.66540034, 18.11700015,
       22.38349858, 15.13675116,  9.02156092, 12.81645757, 17.56760047,
        8.51680373, 25.05214346,  9.04516922, 15.97156262, 11.06644625,
        9.51093817, 20.05181289, 25.04779729, 12.68951782, 22.18259247,
       25.29354906, 12.48575223, 19.01381233, 10.13571448, 15.48670023,
       15.68386313, 12.75724535, 12.90621727,  7.40890802, 11.92464039,
       17.44081721, 17.22366649, 17.5092765 , 12.51023791, 14.56938543,
       20.90825045, 13.27633757,  9.63296388, 10.96998847,  7.01481714,
       13.56604693,  9.33104605, 12.2840357 ,  8.66586146, 13.3213728 ,
       11.09806392, 11.64994657, 11.39164977, 15.30677069, 10.6360893 ,
       18.6643313 , 12.20188468, 17.41913812,  9.41279349, 16.48871777,
       24.09050655,  9.11586534, 10.14967151, 11.11611461, 10.13928522,
       15.06479831])

### Find Model Coefficients i.e metrics to know the performance 

In [21]:
MAE = mean_absolute_error(y_test,test_predictions)

In [22]:
MSE = mean_squared_error(y_test,test_predictions)

In [23]:
RMSE = np.sqrt(MSE)

In [24]:
MAE

0.3595795426686447

In [25]:
MSE

0.22333989542185853

In [26]:
RMSE

0.4725885053848205

### Compare the MAE, MSE and RMSE with Simple Linear output
#### Simple Linear = 1.493775, 3.727928, 1.93078
#### Polynomial Reg= 0.359579, 0.223339, 0.472588

#### We Can clearly see that all the error values are less with Polynomial model

In [28]:
model.coef_

array([ 5.26860919e-02,  1.22591159e-02,  3.93782605e-04, -1.11633478e-04,
        1.11692292e-03, -7.09598343e-05,  1.06877730e-04,  2.73503208e-04,
        1.02517264e-04])

* check the coeff for the corresponding feature and compare which feature or feature combination has more attribute in sales

### How to know which coeff corresponds to which value or feature

In [31]:
# Relate the corresponding value from below
poly_features[0]

array([2.301000e+02, 3.780000e+01, 6.920000e+01, 5.294601e+04,
       8.697780e+03, 1.592292e+04, 1.428840e+03, 2.615760e+03,
       4.788640e+03])

In [32]:
X.iloc[0]

TV           230.1
Radio         37.8
Newspaper     69.2
Name: 0, dtype: float64

#### Let's have look at below observations for more clarity
* first 3 terms are the origial values 230.1, 37.8, 69.2
* 4th term 5.2946 ~ (230.1 ** 2) Polynomial term 
* 5th term 8.6977 ~ 230.1 * 37.8 Interaction term
* 6th term 1.5922 ~ 230.1 * 69.2 Interaction term
* 7th term 1.4288 ~ (37.8 ** 2) Polynomial term
* 8th term 2.6157 ~ 37.8 * 69.2 Interaction term
* 9th term 4.7886 ~ (69.2 ** 2) Polynomial term

# End of the Session